source/ptreferential/ptreferential.cpp

```diff
@@ -275,9 +280,85 @@ std::vector<idx_t> manage_odt_level(const std::vector<type::idx_t>& final_indexe
     return final_indexes;
 }
 
+bool keep_vj(const nt::VehicleJourney* vj,
+             const bt::time_period& period) {
+    if (vj->stop_time_list.empty()) {
+        return false; //no stop time, so it cannot be valid
+    }
+    const auto first_departure_dt = vj->stop_time_list.front().departure_time;
+
+    for (boost::gregorian::day_iterator it(period.begin().date()); it <= period.last().date(); ++it) {
+        if (! vj->validity_pattern->check(*it)) { continue; }
+        bt::ptime vj_dt = bt::ptime(*it, bt::seconds(first_departure_dt));
+        if (period.contains(vj_dt)) { return true; }
+    }
+
+    return false;
+}
+
+std::vector<idx_t> filter_vj_on_period(const std::vector<type::idx_t>& indexes,
+             boost::optional<bt::ptime> since,
+             boost::optional<bt::ptime> until,
+             const type::Data & data) {
+
+    // we create the right period using since, until and the production period
+    if (since && until && until < since) {
+        throw ptref_error("invalid filtering period");
+    }
+    auto start = bt::ptime(data.meta->production_date.begin());
+    auto end = bt::ptime(data.meta->production_date.last(), bt::seconds(24*3600 - 1));
+
+    if (since) {
+        if (data.meta->production_date.is_before(since->date())) {
+            throw ptref_error("invalid filtering period, not in production period");
+        }
+        if (since->date() >= data.meta->production_date.begin()) {
+            start = *since;
+        }
+    }
+    if (until) {
+        if (data.meta->production_date.is_after(until->date())) {
+            throw ptref_error("invalid filtering period, not in production period");
+        }
+        if (until->date() <= data.meta->production_date.last()) {
+            end = *until;
+        }
+    }
+    bt::time_period period {start, end + bt::seconds(1)}; //the end is not in the period, so we add one second
```

Comment:
I though that it shoudl, but i'm not sure, maybe you can ask @stifoon (or it can be changed after)